packages/model-viewer/src/three-components/ARRenderer.ts

```diff
@@ -248,117 +327,330 @@ export class ARRenderer extends EventDispatcher {
-    // which is only added to the session's active input sources immediately
-    // before `selectstart` and removed immediately after `selectend` event.
-    // If we have a 'screen' source here, it means the output canvas was
-    // tapped.
-    const sources = Array.from(session.inputSources)
-                        .filter(input => input.targetRayMode === 'screen');
+    this.dispatchEvent({type: 'modelmove'});
+  }
 
-    if (sources.length === 0) {
+  [$onSelectStart](event: XRInputSourceEvent) {
+    const hitSource = this[$transientHitTestSource];
+    if (hitSource == null) {
       return;
     }
+    const fingers = this[$frame]!.getHitTestResultsForTransientInput(hitSource);
+    const scene = this[$presentedScene]!;
+    const box = this[$placementBox]!;
+
+    if (fingers.length === 1) {
+      this[$inputSource] = event.inputSource;
+      const {axes} = event.inputSource.gamepad;
+
+      const hitPosition = box.getHit(this[$presentedScene]!, axes[0], axes[1]);
+      box.show = true;
+
+      if (hitPosition != null) {
+        this[$isTranslating] = true;
+        this[$lastDragPosition].copy(hitPosition);
+      } else {
+        this[$isRotating] = true;
+        this[$lastScalar] = axes[0];
+      }
+    } else if (fingers.length === 2 && scene.canScale) {
+      box.show = true;
+      this[$isScaling] = true;
+      this[$lastScalar] = this[$fingerSeparation](fingers) / scene.scale.x;
+    }
+  }
+
+  [$onSelectEnd](_event: XRInputSourceEvent) {
+    this[$isTranslating] = false;
+    this[$isRotating] = false;
+    this[$isScaling] = false;
+    this[$inputSource] = null;
+    this[$goalPosition].y +=
+        this[$placementBox]!.offsetHeight * this[$presentedScene]!.scale.x;
+    this[$placementBox]!.show = false
+  }
 
-    const pose = frame.getPose(sources[0].targetRaySpace, this[$refSpace]!);
-    if (pose) {
-      this.placeModel();
+  [$fingerSeparation](fingers: XRTransientInputHitTestResult[]): number {
+    const finger0 = fingers[0].inputSource.gamepad.axes;
+    const finger1 = fingers[1].inputSource.gamepad.axes;
+    const deltaX = finger1[0] - finger0[0];
+    const deltaY = finger1[1] - finger0[1];
+    return Math.sqrt(deltaX * deltaX + deltaY * deltaY);
+  }
+
+  [$processInput](frame: XRFrame) {
+    const hitSource = this[$transientHitTestSource];
+    if (hitSource == null) {
+      return;
+    }
+    if (!this[$isTranslating] && !this[$isScaling] && !this[$isRotating]) {
+      return;
+    }
+    const fingers = frame.getHitTestResultsForTransientInput(hitSource);
+    const scene = this[$presentedScene]!;
+    const scale = scene.scale.x;
+
+    if (this[$isScaling]) {
+      if (fingers.length < 2) {
+        this[$isScaling] = false;
+      } else {
+        const separation = this[$fingerSeparation](fingers);
+        const scale = separation / this[$lastScalar];
+        this[$goalScale] =
+            (scale < SCALE_SNAP && scale > SCALE_SNAP_LOW) ? 1 : scale;
```

Comment:
It seems like `SCALE_SNAP` and `SCALE_SNAP_LOW` represent a boundary, but the naming is strangely asymmetrical if that is the case